src/Microsoft.Restier.Core/ApiBuilderExtensions.cs

```diff
@@ -0,0 +1,165 @@
+
+namespace Microsoft.Restier.Core
+{
+    [CLSCompliant(false)]
+    public static class ApiBuilderExtensions
+    {
+        public static void DefaultInnerMost(IServiceCollection services)
+        {
+            services.AddSingleton<ApiTypeInfo>()
+                .CutoffPrevious<IApiContextFactory, ApiContextFactory>()
+                .CutoffPrevious<IQueryExecutor>(DefaultQueryExecutor.Instance);
+        }
+
+        public static void DefaultOuterMost(IServiceCollection services)
+        {
+            services.ChainPrevious<IApiContextFactory, ApiContextInitializer>();
+            if (!services.HasService<ApiContext>())
+            {
+                services.AddScoped<ContextHolder>()
+                    .AddScoped(sp => sp.GetService<ContextHolder>().Context);
+            }
+        }
+
+        public static Action<IServiceCollection> AttributesConfiguration(Type apiType)
+        {
+            Ensure.NotNull(apiType, "type");
+
+            Action<IServiceCollection> config = null;
+            if (apiType.BaseType != null)
+            {
+                config = AttributesConfiguration(apiType.BaseType);
+            }
+
+            var attributes = apiType.GetCustomAttributes(
+                typeof(IApiConfigurator), false);
+            if (attributes.Length == 0)
+            {
+                return config;
+            }
+
+            config += services =>
+            {
+                foreach (IApiConfigurator e in attributes)
+                {
+                    e.Configure(services, apiType);
+                }
+            };
+            return config;
+        }
+
+        public static ApiBuilder UseAttributes(this ApiBuilder obj, Type apiType)
+        {
+            Ensure.NotNull(obj, "obj");
+
+            return obj.AddInnerTail(AttributesConfiguration(apiType));
+        }
+
+        public static ApiBuilder UseAttributes<TApi>(this ApiBuilder obj)
+            where TApi : class
+        {
+            return obj.UseAttributes(typeof(TApi));
+        }
+
+        public static Action<IServiceCollection> ConventionsConfiguration(Type apiType)
+        {
+            Ensure.NotNull(apiType, "apiType");
+
+            return services =>
+            {
+                ConventionBasedChangeSetAuthorizer.ApplyTo(services, apiType);
+                ConventionBasedChangeSetEntryFilter.ApplyTo(services, apiType);
+                services.CutoffPrevious<IChangeSetEntryValidator, ConventionBasedChangeSetEntryValidator>();
+                ConventionBasedApiModelBuilder.ApplyTo(services, apiType);
+                ConventionBasedOperationProvider.ApplyTo(services, apiType);
+                ConventionBasedEntitySetFilter.ApplyTo(services, apiType);
+            };
+        }
+
+        public static ApiBuilder UseConventions(this ApiBuilder obj, Type apiType)
+        {
+            Ensure.NotNull(obj, "obj");
+
+            return obj.AddOuterMost(ConventionsConfiguration(apiType));
```

Comment:
Conventions default to outer most. Of course consumer could add other configuration call to outer most, after applying conventions, so that override conventions.